src/main/java/edu/kit/compiler/transform/IRExpressionVisitor.java

```diff
@@ -0,0 +1,222 @@
+        Mode resultType = context.getTypeMapper().getDataType(arrayAccessExpressionNode.getResultType()).getMode();
+        return loadFromAddress(arrayAddress, resultType);
+
+    }
+
+    private Node loadFromAddress(Node address, Mode type) {
+        Node mem = getConstruction().getCurrentMem();
+        Node load = getConstruction().newLoad(mem, address, type);
+
+        Node resMem = getConstruction().newProj(load, Mode.getM(), Load.pnM);
+        Node res = getConstruction().newProj(load, type, Load.pnRes);
+        getConstruction().setCurrentMem(resMem);
+        return res;
+    }
+
+    @Override
+    public Node visit(IdentifierExpressionNode identifierExpressionNode) {
+        Construction con = context.getConstruction();
+        if (identifierExpressionNode.getDefinition().getKind() == DefinitionKind.LocalVariable) {
+            Mode mode = getMode(identifierExpressionNode.getResultType());
+            return con.getVariable(context.getVariableIndex(identifierExpressionNode.getIdentifier()), mode);
+        } else if (identifierExpressionNode.getDefinition().getKind() == DefinitionKind.Parameter) {
+            return context.createParamNode(identifierExpressionNode.getIdentifier());
+        } else {
+            throw new IllegalArgumentException();
+        }
+    }
+
+    @Override
+    public Node visit(ThisExpressionNode thisExpressionNode) {
+        return context.createThisNode();
+    }
+
+    @Override
+    public Node visit(ValueExpressionNode valueExpressionNode) {
+        // determine type of value expression and value and set the expression
+        return switch (valueExpressionNode.getValueType()) {
+            case Null -> handleNullExpressionNode();
+            case False, True -> IRBooleanExpressions.asValue(context, valueExpressionNode);
+            case IntegerLiteral -> handleIntegerExpressionNode(valueExpressionNode);
+        };
+    }
+
+    private Node handleIntegerExpressionNode(ValueExpressionNode valueExpressionNode) {
+        TargetValue tar_val = new TargetValue(valueExpressionNode.getLiteralValue().get().intValue(), Mode.getIs());
+        return getConstruction().newConst(tar_val);
+    }
+
+    private Node handleNullExpressionNode() {
+        return getConstruction().newAddress(new Entity(Pointer.NULL));
+    }
+
+    @Override
+    public Node visit(NewObjectExpressionNode newObjectExpressionNode) {
+        TypeMapper.ClassEntry classNode = context.getTypeMapper().getClassEntry(newObjectExpressionNode.getTypeName());
+        int size = classNode.getSize();
+        return callCalloc(1, size, classNode.getClassType());
+    }
+
+    @Override
+    public Node visit(NewArrayExpressionNode newArrayExpressionNode) {
+        Type dataType = context.getTypeMapper().getDataType(newArrayExpressionNode.getElementType());
+        Node arrayLength = newArrayExpressionNode.getLength().accept(this);
+        return callCalloc(arrayLength, dataType.getSize(), dataType);
```

Comment:
As for new object, the `dataType` is not the correct result type. You need a pointer to an unsized array of `dataType` (this is also provided by `TypeMapper`)